File: remove_dup_linkedlist.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution(object):
    def deleteDuplicates(self, head):
        """
        :type head: ListNode
        :rtype: ListNode
        """
        if head==None or head.next==None:
            return head
        cur=head
        newhead=None
        newcur=None
        distinct=True
        first=True
        curvalue=cur.val
        while cur!=None:
            curvalue=cur.val
            if cur.next is None or cur.next.val!=curvalue:
                
                if distinct:
                    if first:
                        newhead=ListNode(curvalue)
                        newcur=newhead
                        
                        first=False
                    else:
                        newcur.next=ListNode(curvalue)
                        newcur=newcur.next
                
                    
                distinct=True
            else:
                distinct=False
            cur=cur.next
        return newhead


        

File: remove_dup_linkedlist.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */
public class Solution {
    public ListNode deleteDuplicates(ListNode head) {
        if (head==null || head.next==null) return head;
        ListNode current =head;
        ListNode previous=head;
        
        Boolean cur_dup=false;
        while (current!=null){
            while (current.next!=null && (current.next.val==current.val)){
                cur_dup=true;
                current.next=current.next.next;
                previous=null;
            }
            if (cur_dup){
                cur_dup=false;
                if (previous==null){
                    if (current.next==null){
                        head=null;
                    }
                    else{
                        head=current.next.next;
                        current=head;
                    }
                
                }
                else{
                    previous.next=current.next.next;
                    current=previous.next;
                    previous=current;
                }
               
            }
            else{
                previous=current;
                current=current.next;
            }
            
        }
        return head;
    }
    
}